### PR TITLE
Add recipe for Gensim

### PR DESCRIPTION
**Pull request by souravsingh**
Adds a new recipe for Gensim.

Ping @tmylk @menshikh-iv

### CODE REVIEW

File: recipes/gensim/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    - numpy >=1.3
+    - scipy >=0.7.0
+    - six >=1.5.0
+    - smart_open >=1.2.1
+
+test:
+  requires:
+    - nose
+    - testfixtures
+    - unittest
```
Comment (mariusvniekerk) [inline recipes/gensim/meta.yaml]: unittest is a builtin

File: recipes/gensim/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - numpy >=1.3
```
Comment (mariusvniekerk) [inline recipes/gensim/meta.yaml]: Since gensim uses cython, i would change this to 

```
numpy x.x
cython
``` 
and change the numpy section in the run area too.

This will build you a version of gensim that is compatible with the numpy ABI for the versions that conda forge supports

**Comment by mariusvniekerk** [comment]
So generally we don't run the unit test suite of a library as part of conda forge packaging.  Instead we validate that things are importable and correctly linked.  Running unit tests is typically done upstream in the project itself.

File: recipes/gensim/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    - numpy x.x
+    - cython
+    - scipy >=0.7.0
+    - six >=1.5.0
+    - smart_open >=1.2.1
+    
+  run:
+    - python
+    - numpy x.x
+    - cython
```
Comment (mariusvniekerk) [inline recipes/gensim/meta.yaml]: cython is probably not needed as a runtime dependency

File: recipes/gensim/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - numpy x.x
+    - cython
+    - scipy >=0.7.0
```
Comment (menshikh-iv) [inline recipes/gensim/meta.yaml]: Please use numpy/scipy versions from `develop` branch

File: recipes/gensim/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    - python
+    - numpy x.x
+    - cython
+    - scipy >=0.7.0
+    - six >=1.5.0
+    - smart_open >=1.2.1
+    
+test:
+  requires:
+    - nose
```
Comment (mariusvniekerk) [inline recipes/gensim/meta.yaml]: Its probably sufficient initially to have

```test:
    imports:
        - gensim
        - gensim.foo
```
for all the standard importable modules

**Comment by souravsingh** [comment]
@mariusvniekerk @menshikh-iv The tests on the recipe have passed, but I remember @tmylk saying that the whole test suite must be run. Should we go ahead with running the import tests?

**Comment by souravsingh** [comment]
Ping @menshikh-iv

**Comment by tmylk** [comment]
If the gensim suite tests of the assembled conda recipe fail then the package shouldn't be released.

**Comment by souravsingh** [comment]
Thanks @tmylk. I will be replacing the import tests with the full gensim tests.

**Comment by souravsingh** [comment]
@menshikh-iv The test suite is not running due to the unittest2 import. Do we replace all the ``unittest2`` imports with unittest?

File: recipes/gensim/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - numpy >=1.11.3
+    - cython
```
Comment (piskvorky) [inline recipes/gensim/meta.yaml]: Why is this needed? I don't think `cython` is a dependency of gensim.

File: recipes/gensim/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  commands:
+    - nosetests --exe -v gensim
+    
+about:
+  home: http://github.com/RaRe-Technologies/gensim
+  license: LGPL 3.0
+  license_file: COPYING
+  license_family: LGPL
+  summary: 'A library for topic modelling and document indexing'
+  description: |
```
Comment (piskvorky) [inline recipes/gensim/meta.yaml]: Please use the official tagline and description (see https://github.com/RaRe-Technologies/gensim).

File: recipes/gensim/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    
+  commands:
+    - nosetests --exe -v gensim
+    
+about:
+  home: http://github.com/RaRe-Technologies/gensim
+  license: LGPL 3.0
+  license_file: COPYING
+  license_family: LGPL
+  summary: 'Topic Modelling in Python'
```
Comment (piskvorky) [inline recipes/gensim/meta.yaml]: Tagline = `Topic Modelling for Humans`.

File: recipes/gensim/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  description: |
+    Gensim is a Python library for topic modelling, document indexing 
+    and similarity retrieval with large corpora. 
+    Target audience is the natural language processing(NLP)
```
Comment (piskvorky) [inline recipes/gensim/meta.yaml]: Missing space before opening bracket.

File: recipes/gensim/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  commands:
+    - nosetests --exe -v gensim
+    
+about:
+  home: http://github.com/RaRe-Technologies/gensim
+  license: LGPL 3.0
+  license_file: COPYING
+  license_family: LGPL
+  summary: 'Topic Modelling in Python'
+  description: |
```
Comment (piskvorky) [inline recipes/gensim/meta.yaml]: The `|` is some specific formatter?

Comment (souravsingh) [inline recipes/gensim/meta.yaml]: Yes

**Comment by souravsingh** [comment]
@menshikh-iv The tests in Travis CI are failing due to build timeout.

**Comment by menshikh-iv** [comment]
@souravsingh looks like slow version of w2v/d2v, please check `FAST_VERSION` `python -c "from gensim.models.word2vec import FAST_VERSION; print(FAST_VERSION)"`, should be `>-1`

**Comment by souravsingh** [comment]
@menshikh-iv The command prints ``1`` across all builds. So I believe the build is using the ``FAST_VERSION`` of ``doc2vec``.

**Comment by piskvorky** [comment]
Is this something we want to eventually merge, or do conda-force recipes operate on updating PRs continuously?

**Comment by souravsingh** [comment]
@piskvorky We are looking to merge this recipe, but due to problems running the full ``nosetests`` test-suite on the recipe, the PR is getting stalled. We want the PR to fulfill all CI tests so the recipe can work across all platforms without problem.

**Comment by souravsingh** [comment]
@piskvorky The recipe passes all the tests. Would you like to take a look and and any reviews?

**Comment by piskvorky** [comment]
I'll leave it to @menshikh-iv , I have no experience with conda.

File: recipes/gensim/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,74 @@
+{% set name = "gensim" %}
+{% set version = "2.3.0" %}
+{% set sha256 = "7d0dccc7d2c576e270037949874800b7cfbc86ef081ff981483f612cd18e223f" %}
```
Comment (menshikh-iv) [inline recipes/gensim/meta.yaml]: Should we update this every release?

File: recipes/gensim/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    - gensim.models.wrappers
+    - gensim.parsing
+    - gensim.scripts
+    - gensim.similarities
+    - gensim.sklearn_integration
+    - gensim.summarization
+    - gensim.test
+    - gensim.topic_coherence
+    
+  commands:
```
Comment (menshikh-iv) [inline recipes/gensim/meta.yaml]: Why you removed test run?

File: recipes/gensim/meta.yaml (hunk 11 above)
Comment (souravsingh) [inline recipes/gensim/meta.yaml]: The version and the sha256 checksum is updated for every release.

File: recipes/gensim/meta.yaml (hunk 12 above)
Comment (souravsingh) [inline recipes/gensim/meta.yaml]: There are a few reasons for removing the test runs-
1) Tensorflow doesn't support 32-bit Windows, so running Gensim tests for win-32 on Appveyor will fail. It is possible to skip win-32 builds, but it might defeat the purpose of full support.

2) The test run times out in Travis( which is used for building recipe for OSX) at Doc2Vec parallel training test due to no output for 10 mins.

File: recipes/gensim/meta.yaml (hunk 11 above)
Comment (menshikh-iv) [inline recipes/gensim/meta.yaml]: Should we update it manually every release?

File: recipes/gensim/meta.yaml (hunk 12 above)
Comment (menshikh-iv) [inline recipes/gensim/meta.yaml]: Yes, we want to run tests for sure

File: recipes/gensim/meta.yaml (hunk 11 above)
Comment (souravsingh) [inline recipes/gensim/meta.yaml]: Yes, We only need to update the ``version`` and the ``sha256`` for every release. This is done by sending a PR to the complete gensim feedstock repository.

File: recipes/gensim/meta.yaml (hunk 12 above)
Comment (souravsingh) [inline recipes/gensim/meta.yaml]: Tests are failing for ``sklearn_integration``
```
======================================================================
ERROR: testPipeline (gensim.test.test_sklearn_integration.TestSklLdaModelWrapper)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/staged-recipes/build_artefacts/gensim_1503413959047/_t_env/lib/python2.7/site-packages/gensim/test/test_sklearn_integration.py", line 153, in testPipeline
    text_lda.fit(corpus, data.target)
  File "/staged-recipes/build_artefacts/gensim_1503413959047/_t_env/lib/python2.7/site-packages/sklearn/pipeline.py", line 257, in fit
    Xt, fit_params = self._fit(X, y, **fit_params)
  File "/staged-recipes/build_artefacts/gensim_1503413959047/_t_env/lib/python2.7/site-packages/sklearn/pipeline.py", line 226, in _fit
    self.steps[step_idx] = (name, fitted_transformer)
TypeError: 'tuple' object does not support item assignment
-------------------- >> begin captured logging << --------------------
```

Comment (menshikh-iv) [inline recipes/gensim/meta.yaml]: This is a problem with `sklearn==0.19.0`, we already fix it in `develop` branch. Now you can "hardcode" sklearn to `0.18.2` to avoid this (and remove in next gensim release)

File: recipes/gensim/meta.yaml (hunk 11 above)
Comment (menshikh-iv) [inline recipes/gensim/meta.yaml]: Sorry, I don't catch, should we to update this two lines over creating PR here after each release?

Comment (souravsingh) [inline recipes/gensim/meta.yaml]: If a new release has been made when this recipe hasn't been merged, we update it the two lines from here to ensure that the latest version is packaged.

Otherwise, if this patch is merged before the new release, we will be sending a pull request to the feedstock repository of gensim, which would look similar to this- https://github.com/conda-forge/validictory-feedstock and the maintainers of the recipe(mentioned in the recipe-maintainers section of the recipe) will take a look and merge the patch to update the version of the conda package.

The ``sha256`` comes from the project link in [pypi.io](https://pypi.org/project/gensim/#files).

File: recipes/gensim/meta.yaml (hunk 12 above)
Comment (souravsingh) [inline recipes/gensim/meta.yaml]: The pinning has solved the ``sklearn`` tests, but there is a new failure-
```
======================================================================
ERROR: Test Keras 'Embedding' layer returned by 'get_embedding_layer' function for a smaller version of the 20NewsGroup classification problem.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/site-packages/gensim/test/test_keras_integration.py", line 100, in testEmbeddingLayer20NewsGroup
    data = fetch_20newsgroups(subset='train', categories=['alt.atheism', 'comp.graphics', 'sci.space'])
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/site-packages/sklearn/datasets/twenty_newsgroups.py", line 225, in fetch_20newsgroups
    cache_path=cache_path)
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/site-packages/sklearn/datasets/twenty_newsgroups.py", line 91, in download_20newsgroups
    opener = urlopen(URL)
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/urllib2.py", line 1228, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/staged-recipes/build_artefacts/gensim_1503567059210/_t_env/lib/python2.7/urllib2.py", line 1198, in do_open
    raise URLError(err)
URLError: <urlopen error [Errno 110] Connection timed out>
```

**Comment by souravsingh** [comment]
@piskvorky @menshikh-iv I have fixed the recipe and it should now work with ``linux-64`` and ``win-64`` platform. There is a failure with Python 2.7 build due to an HTTPerror, which I believe is sporadic and should be fixed by a simple build restart. I also suggest skipping support for OSX until we can figure out why the build times out.

@ocefpaf Is it possible to restart Appveyor builds for the recipe? The Python 2.7 build has failed due to a ``ConnectionError``

**Comment by ocefpaf** [comment]
Done. But you can also do that by closing and opening the PR.

**Comment by souravsingh** [comment]
@piskvorky The recipe is now ready for deployment as a feedstock. The recipe currently supports Linux and Windows 64-bit, since it might create problems for Windows 32-bit if they wish to run the Gensim for a larger corpus. 

I have skipped support for OSX for now since Travis times out everytime when we run the ``nosetests`` command.

**Comment by piskvorky** [comment]
Thanks, but I don't have merge rights in this repo. This is for @menshikh-iv -- what is the merging process here?

**Comment by souravsingh** [comment]
@piskvorky The PR will be merged by one of the reviewers from conda-forge organization(Filipe handles most of the merging stuff). Once the recipe is merged, The ``conda-smithy`` will start creating a feedstock repository for the recipe, something similar to this repository- https://github.com/conda-forge/tensorflow-feedstock and the built tarballs will be uploaded to the ``conda-forge`` channel, from where the users can download the tarball directly.

**Comment by mariusvniekerk** [comment]
@ocefpaf @ericdill @jakirkham Is this ready to go?

**Comment by ocefpaf** [comment]
We are having some trouble in `master` with the feedstock creation. We'll resume merging PRs as soon as that is solved. (I'll look into the problem soon.)

**Comment by menshikh-iv** [comment]
Ignoring MacOS looks suspicious, but we can resolve it later.
LGTM for me, good job @souravsingh:+1:

**Comment by souravsingh** [comment]
@ocefpaf Is the recipe ready to merge?

**Comment by souravsingh** [comment]
@ocefpaf The Gensim doesn't support Python 3.4. It only supports Python 2.7, 3.5 and 3.6

**Comment by ocefpaf** [comment]
> @ocefpaf The Gensim doesn't support Python 3.4. It only supports Python 2.7, 3.5 and 3.6

Don't worry about the pre-selector there. Conda-forge does not build Python 3.4. (I added it for reasons :wink:)

However, if py34 is not really support you need to add a skip statement.

**Comment by menshikh-iv** [comment]
> @ocefpaf The Gensim doesn't support Python 3.4. It only supports Python 2.7, 3.5 and 3.6
Incorrect. Gensim supports all versions `>= 2.7`.

**Comment by ocefpaf** [comment]
> Incorrect. Gensim supports all versions >= 2.7.

Thanks @menshikh-iv. Good to know!

**Comment by souravsingh** [comment]
My bad, I thought Gensim doesn't support Python 3.4 since there is not build matrix for py3.4 in Travis.
